src/array.c

```diff
@@ -399,14 +411,47 @@ const void * const*array_get_buffer(Array *ar)
  */
 enum cc_stat array_index_of(Array *ar, void *element, size_t *index)
 {
-    size_t i;
-    for (i = 0; i < ar->size; i++) {
-        if (ar->buffer[i] == element) {
-            *index = i;
-            return CC_OK;
+    if(ar->is_sorted){
+        size_t left=0, right = ar->size-1, middle;
+        int cmp, lastcmp = -1;
+        bool done=false;
+        while(!done){
+            middle = (right-left)/2 + left;
+	    middle = (right-left == 1) && lastcmp > 0 ? left + 1 : middle;
+            cmp = ar->cmp(element,ar->buffer[middle]);
+            if(cmp < 0){
+                right = middle;
+            }else if(cmp > 0){
+                left = middle;
+            }else if(cmp == 0){
+                done = true;
+            }
+            if(left > right) return CC_ERR_OUT_OF_RANGE;
+	    lastcmp = cmp;
         }
```

Comment:
It still doesn't seem to work. I've tried it with an array of `1, 4, 8, 15, 123` and it gets stuck while trying to find `5`. Basically it just ping-pongs between indices 1 and 2 tying to find the value that should be between them. I suggest that we just go with the standard implementation instead of trying to patch this one.  This is one of those deceptively simple algorithms that are very easy to get wrong.

I suggest we do something like this:

Put bsearch in a function so that we don't have to complicate loop exit conditions
```c
static enum cc_stat array_bsearch(Array *ar, void *key, size_t *index)
{
    size_t left  = 0;
    size_t right = ar->size - 1;

    while (left <= right) {
        size_t middle = (left + right) >> 1;
        int cmp = ar->cmp(key, ar->buffer[middle]);

        if (cmp < 0) {
            right = middle - 1;
        } else if (cmp > 0) {
            left = middle + 1;
        } else {
            *index = middle;
            return CC_OK;
        }
    }
    return CC_ERR_OUT_OF_RANGE;
}
```

then in array_index_of:
```c
...
if (ar->is_sorted) {
    if (array_bsearch(ar, element, index) != CC_OK)
        return CC_ERR_OUT_OF_RANGE;

    while (*index > 0) {
         ...
    } 
    ...
}
```